### PR TITLE
Allow blocks to be retained after compaction

### PR DESCRIPTION
**Pull request by darshanime**
Signed-off-by: darshanime <deathbullet@gmail.com>

Add ability to retain blocks after compaction. 
Fixes https://github.com/thanos-io/thanos/issues/2344

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: cmd/thanos/compact.go [hunk 1]
```diff
@@ -149,6 +149,8 @@ func registerCompact(m map[string]setupFunc, app *kingpin.Application) {
 	webPrefixHeaderName := cmd.Flag("web.prefix-header", "Name of HTTP request header used for dynamic prefixing of UI links and redirects. This option is ignored if web.external-prefix argument is set. Security risk: enable this option only if a reverse proxy in front of thanos is resetting the header. The --web.prefix-header=X-Forwarded-Prefix option can be useful, for example, if Thanos UI is served via Traefik reverse proxy with PathPrefixStrip option enabled, which sends the stripped prefix value in X-Forwarded-Prefix header. This allows thanos UI to be served on a sub-path.").Default("").String()
 	label := cmd.Flag("bucket-web-label", "Prometheus label to use as timeline title in the bucket web UI").String()
 
+	retainCompactedBlocks := cmd.Flag("retain-compacted-blocks", "Do not remove old blocks even after they are safely compacted, but move them to another directory").Default("false").Bool()
```
Comment (bwplotka) [inline cmd/thanos/compact.go:152]: Cool, we were thinking about this, but actually we though about separate bucket :thinking: 
Thoughts?

Essentially backup mode

**Comment by belm0** [comment]
document the use case in the compactor docs?

**Comment by darshanime** [comment]
change to retain in separate bucket ✔️ 
added a note about usecase in docs ✔️

File: docs/components/compact.md [hunk 2]
```diff
@@ -49,6 +49,8 @@ Ideally, you will have equal retention set (or no retention at all) to all resol
 
 Not setting this flag, or setting it to `0d`, i.e. `--retention.resolution-X=0d`, will mean that samples at the `X` resolution level will be kept forever.
 
+Optionally, you can also retain blocks which are compacted into larger blocks. You can supply `--objstore-retain-compacted-blocks-to-bucket.config` and Thanos will move compacted blocks to the specified bucket. This can be useful if you do not want to lose blocks during rewrites, which can happen in compaction, delete series etc.
```
Comment (belm0) [inline docs/components/compact.md:52]: it's not clear-- does this apply to only the original blocks, or also to higher-order compactions?

if the goal is to keep a backup of the original prometheus blocks, it seems the latter is not necessary

**Comment by bwplotka** [comment]
Thanks for this but I think we need to have larger discussion first.

For me backing up all of the blocks might be not efficient - we are actually replicate tons of things for backup purposes, which is just not needed. 

The perfect way TBH would be to use `thanos bucket replicate` to always backup 2h blocks to some cold storage if you want an overall backup.

If you want backup only for retention then just stating NO retention and adding some `min` time for compactor might make more sense.

What's the use case for you? I would say it's a perfect item to discuss on our Thanos public community meeting if you want to join: https://thanos.io/community.md/

**Comment by darshanime** [comment]
Hi @bwplotka 👋 

To be honest, I am not an end user of Thanos, but am trying to contribute to Thanos so as to participate and learn more about building distributed systems. 

I came across the issue #2344 and thought I'd try to implement it since it relates to `delete_series` (which is a project I am hoping to get selected for in Community Bridge), . 
I should have asked for more details/prepared a document to get concensus before working on it, will keep that in mind going forward 👍

**Comment by bwplotka** [comment]
Thanks for this.

It's a good pattern to first describe quickly how you plan to do it so we can quickly discuss it.

Anyway, this is only my opinion, maybe it's fine to duplicate backup... Some opinions from other @thanos-io/thanos-maintainers would be nice :hugs:

**Comment by brancz** [comment]
I think in particular because we can never know when compaction happens it's probably safest to put backup logic into it, even if that somewhat overlaps with the replication tooling we provide (as those don't have the same safety guarantees as the compactor backup process would be able to provide).

**Comment by darshanime** [comment]
Hi, can this be closed? 
Agree a better PR can be made after getting consensus on what the best retention strategy would be.

**Comment by kakkoyun** [comment]
Closing this as @darshanime requested.

**Pull request closed without merging**